{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Biopython Annotated Chromosomes Diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from collections import OrderedDict\n",
    "import dash_bootstrap_components as dbc\n",
    "from dash import Dash, dcc, html, Input, Output, State, dash_table, callback_context\n",
    "\n",
    "import csv\n",
    "\n",
    "from flask import Flask\n",
    "import sqlite3\n",
    "import numpy as np\n",
    "import plotly.graph_objs as go\n",
    "\n",
    "import os\n",
    "import zlib\n",
    "import subprocess\n",
    "import zlib\n",
    "import uuid\n",
    "\n",
    "from collections import OrderedDict\n",
    "\n",
    "import dash\n",
    "from dash import no_update # https://community.plotly.com/t/error-expected-the-output-type-to-be-a-list-or-tuple-but-got-none/34744/6\n",
    "from flask import Flask, send_from_directory\n",
    "from urllib.parse import quote as urlquote\n",
    "\n",
    "from io import StringIO\n",
    "from Bio import SeqIO\n",
    "from Bio import Phylo\n",
    "from Bio import AlignIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# import numpy as np\n",
    "from io import BytesIO\n",
    "import base64\n",
    "\n",
    "\n",
    "# import pandas as pd\n",
    "from plotly.tools import mpl_to_plotly\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Rectangle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Use this function to generate the chromosome.tsv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_list = []\n",
    "for file_name in os.listdir('./'):\n",
    "    suffix = file_name.split(\".\")[-1]\n",
    "    if file_name[-4:] != \"a.gz\":\n",
    "        continue\n",
    "    if \"Athaliana\" in file_name:\n",
    "        genotype_id = \"Arabidopsis\"\n",
    "    else:\n",
    "        genotype_id = \"B73v4\"\n",
    "\n",
    "    fasta_file = gzip.open(file_name, mode='rt')\n",
    "    for chrom in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        # Check if the sequence ID could be coerced into an integer, which is likely to be a chromosome\n",
    "        if not any([chrom.id.replace(prefix, \"\").isdigit() for prefix in [\"Chr\", \"chr\", \"C\", \"c\"]]):\n",
    "            continue\n",
    "        # Add species and genotype columns for a better annotation\n",
    "        chrom_list.append([file_name, genotype_id, chrom.id, len(chrom.seq)])\n",
    "\n",
    "# Dataframe that contains each chromosome lenght\n",
    "df = pd.DataFrame(chrom_list)\n",
    "df.columns = [\"genome\", \"genotype_id\",\"chr\", \"len\"]\n",
    "df.to_csv(\"chromosome.tsv\", index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome</th>\n",
       "      <th>genotype_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Athaliana_167_TAIR10.fa.gz</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>30427671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Athaliana_167_TAIR10.fa.gz</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19698289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Athaliana_167_TAIR10.fa.gz</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr3</td>\n",
       "      <td>23459830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Athaliana_167_TAIR10.fa.gz</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr4</td>\n",
       "      <td>18585056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Athaliana_167_TAIR10.fa.gz</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr5</td>\n",
       "      <td>26975502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Zmays_493_APGv4.fa.gz</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>1</td>\n",
       "      <td>307041717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Zmays_493_APGv4.fa.gz</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>2</td>\n",
       "      <td>244442276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Zmays_493_APGv4.fa.gz</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>3</td>\n",
       "      <td>235667834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Zmays_493_APGv4.fa.gz</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>4</td>\n",
       "      <td>246994605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Zmays_493_APGv4.fa.gz</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>5</td>\n",
       "      <td>223902240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Zmays_493_APGv4.fa.gz</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>6</td>\n",
       "      <td>174033170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Zmays_493_APGv4.fa.gz</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>7</td>\n",
       "      <td>182381542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Zmays_493_APGv4.fa.gz</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>8</td>\n",
       "      <td>181122637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Zmays_493_APGv4.fa.gz</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>9</td>\n",
       "      <td>159769782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Zmays_493_APGv4.fa.gz</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>10</td>\n",
       "      <td>150982314</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        genome  genotype_id   chr        len\n",
       "0   Athaliana_167_TAIR10.fa.gz  Arabidopsis  Chr1   30427671\n",
       "1   Athaliana_167_TAIR10.fa.gz  Arabidopsis  Chr2   19698289\n",
       "2   Athaliana_167_TAIR10.fa.gz  Arabidopsis  Chr3   23459830\n",
       "3   Athaliana_167_TAIR10.fa.gz  Arabidopsis  Chr4   18585056\n",
       "4   Athaliana_167_TAIR10.fa.gz  Arabidopsis  Chr5   26975502\n",
       "5        Zmays_493_APGv4.fa.gz        B73v4     1  307041717\n",
       "6        Zmays_493_APGv4.fa.gz        B73v4     2  244442276\n",
       "7        Zmays_493_APGv4.fa.gz        B73v4     3  235667834\n",
       "8        Zmays_493_APGv4.fa.gz        B73v4     4  246994605\n",
       "9        Zmays_493_APGv4.fa.gz        B73v4     5  223902240\n",
       "10       Zmays_493_APGv4.fa.gz        B73v4     6  174033170\n",
       "11       Zmays_493_APGv4.fa.gz        B73v4     7  182381542\n",
       "12       Zmays_493_APGv4.fa.gz        B73v4     8  181122637\n",
       "13       Zmays_493_APGv4.fa.gz        B73v4     9  159769782\n",
       "14       Zmays_493_APGv4.fa.gz        B73v4    10  150982314"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"chromosome.tsv\", sep=\"\\t\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = [\"Zm00001d020673\",\n",
    "    \"Zm00001d020562\",\n",
    "    \"Zm00001d020177\",\n",
    "    \"Zm00001d024875\",\n",
    "    \"Zm00001d006622\",\n",
    "    \"Zm00001d025200\",\n",
    "    \"Zm00001d002610\",\n",
    "    \"Zm00001d024515\",\n",
    "    \"Zm00001d007918\",\n",
    "    \"Zm00001d034095\",\n",
    "    \"Zm00001d007112\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "con = sqlite3.connect(\"../SQNce.db\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_coordinates(genotype, gene_id):\n",
    "    con = sqlite3.connect(\"../SQNce.db\") # deploy with this\n",
    "    cursorObj = con.cursor()\n",
    "    df = pd.read_sql_query('''SELECT * \n",
    "                    FROM gene_coordinates \n",
    "                    WHERE genotype_id = \"{0}\"\n",
    "                    AND gene_id = \"{1}\"\n",
    "                    '''.format(genotype, gene_id), con)\n",
    "    return(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>genotype_id</th>\n",
       "      <th>gene_chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_orientation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d020673</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>7</td>\n",
       "      <td>127586707</td>\n",
       "      <td>127588589</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d020562</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>7</td>\n",
       "      <td>122351840</td>\n",
       "      <td>122352624</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d020177</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>7</td>\n",
       "      <td>97680746</td>\n",
       "      <td>97682402</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d024875</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>10</td>\n",
       "      <td>91796481</td>\n",
       "      <td>91798115</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d006622</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>2</td>\n",
       "      <td>212780800</td>\n",
       "      <td>212783824</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d025200</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>10</td>\n",
       "      <td>108836325</td>\n",
       "      <td>108838302</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d002610</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>2</td>\n",
       "      <td>16708568</td>\n",
       "      <td>16710668</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d024515</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>10</td>\n",
       "      <td>75531512</td>\n",
       "      <td>75533071</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d007918</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>2</td>\n",
       "      <td>242408695</td>\n",
       "      <td>242410359</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d034095</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>1</td>\n",
       "      <td>283893649</td>\n",
       "      <td>283895468</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d007112</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>2</td>\n",
       "      <td>222488036</td>\n",
       "      <td>222489811</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d007112</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>2</td>\n",
       "      <td>222488036</td>\n",
       "      <td>222489811</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_id genotype_id gene_chr  gene_start   gene_end gene_orientation\n",
       "0  Zm00001d020673       B73v4        7   127586707  127588589                -\n",
       "0  Zm00001d020562       B73v4        7   122351840  122352624                -\n",
       "0  Zm00001d020177       B73v4        7    97680746   97682402                +\n",
       "0  Zm00001d024875       B73v4       10    91796481   91798115                +\n",
       "0  Zm00001d006622       B73v4        2   212780800  212783824                +\n",
       "0  Zm00001d025200       B73v4       10   108836325  108838302                +\n",
       "0  Zm00001d002610       B73v4        2    16708568   16710668                +\n",
       "0  Zm00001d024515       B73v4       10    75531512   75533071                -\n",
       "0  Zm00001d007918       B73v4        2   242408695  242410359                +\n",
       "0  Zm00001d034095       B73v4        1   283893649  283895468                +\n",
       "0  Zm00001d007112       B73v4        2   222488036  222489811                +\n",
       "0  Zm00001d007112       B73v4        2   222488036  222489811                +"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype = \"B73v4\"\n",
    "coords = pd.DataFrame()\n",
    "for gene in gene_list:\n",
    "    coords = pd.concat([df, get_gene_coordinates(genotype, gene)])\n",
    "coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(212780800, 212780800, None, 'Zm00001d006622', 'black'),\n",
       " (16708568, 16708568, None, 'Zm00001d002610', 'black'),\n",
       " (242408695, 242408695, None, 'Zm00001d007918', 'black'),\n",
       " (222488036, 222488036, None, 'Zm00001d007112', 'black'),\n",
       " (222488036, 222488036, None, 'Zm00001d007112', 'black')]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features = []\n",
    "for row in coords[coords[\"gene_chr\"]==str(2)].values:\n",
    "    features.append((row[3], row[3], None, row[0], \"black\",))\n",
    "features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "from reportlab.lib.units import cm\n",
    "from Bio import SeqIO\n",
    "from Bio.Graphics import BasicChromosome\n",
    "\n",
    "selected = df[df[\"genotype_id\"]==\"B73v4\"]\n",
    "entries = selected[[\"chr\", \"len\"]].values.tolist()\n",
    "max_len = selected[\"len\"].max()\n",
    "telomere_length = 0\n",
    "\n",
    "chr_diagram = BasicChromosome.Organism()\n",
    "chr_diagram.page_size = (29.7 * cm, 21 * cm)  # A4 landscape\n",
    "\n",
    "for index, (name, length) in enumerate(entries):\n",
    "    \n",
    "    # For each chrosome generate a feature list of tuples (start, start, strand, name, color,)\n",
    "    features = []\n",
    "    for row in coords[coords[\"gene_chr\"]==name].values:\n",
    "        features.append((row[3], row[3], None, row[0], \"black\",))\n",
    "    \n",
    "    cur_chromosome = BasicChromosome.Chromosome(name)\n",
    "    # Set the scale to the MAXIMUM length plus the two telomeres in bp,\n",
    "    # want the same scale used on all five chromosomes so they can be\n",
    "    # compared to each other\n",
    "    cur_chromosome.scale_num = max_len + 2\n",
    "\n",
    "    # Record an Artemis style integer color in the feature's qualifiers,\n",
    "    # 1 = Black, 2 = Red, 3 = Green, 4 = blue, 5 =cyan, 6 = purple\n",
    "    \n",
    "    # The features can either be SeqFeature objects, or tuples of values: start (int), \n",
    "    # end (int), strand (+1, -1, O or None), label (string), ReportLab color (string or object), \n",
    "    # and optional ReportLab fill color.\n",
    "    \n",
    "    # This will it to every chromsome. Make chromsome specific lists\n",
    "    \n",
    "    \n",
    "    # Add an opening telomere\n",
    "    start = BasicChromosome.TelomereSegment()\n",
    "    start.scale = telomere_length\n",
    "    cur_chromosome.add(start)\n",
    "    \n",
    "    # Add a body - again using bp as the scale length here.\n",
    "    body = BasicChromosome.AnnotatedChromosomeSegment(length, features)\n",
    "    body.scale = length\n",
    "    cur_chromosome.add(body)\n",
    "    \n",
    "    # Add a closing telomere\n",
    "    end = BasicChromosome.TelomereSegment(inverted=True)\n",
    "    end.scale = telomere_length\n",
    "    cur_chromosome.add(end)\n",
    "    \n",
    "    # This chromosome is done\n",
    "    chr_diagram.add(cur_chromosome)\n",
    "\n",
    "    from reportlab.graphics import renderPM\n",
    "reportlab = chr_diagram.draw(None, \"Genome Diagram\")\n",
    "renderPM.drawToFile(reportlab, 'example1.png', 'PNG')\n",
    "#chr_diagram.draw(\"test_chrom.pdf\", \"Genome Diagram\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<reportlab.graphics.shapes.Drawing at 0x29117d311c0>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_diagram.draw(None, \"Genome Diagram\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
